sbi/utils/conditional_density.py

```diff
@@ -357,15 +359,54 @@ def __init__(self, MDN_Posterior: DirectPosterior):
             self.__dict__ = deepcopy(MDN_Posterior).__dict__
 
             # MoG parameters
-            self.S = None
-            self.m = None
-            self.mc = None
+            self.precs = None
+            self.means = None
+            self.logits = None
+            self.sumlogdiag = None
             self.support = self._prior.support
 
-            self.extract_mixture_components()
+            self.extract_and_transform_mog()
 
         else:
             raise AttributeError("Posterior does not contain a MDN.")
+            
+    def extract_and_transform_mog(
+        self, context: Tensor = None
+    ) -> Tuple[Tensor, Tensor, Tensor]:
+        """Extracts the Mixture of Gaussians (MoG) parameters
+        from the MDN at either the default x or input x.
+
+        Args:
+            x: x at which to evaluate the MDN in order
+                to extract the MoG parameters.
+        """
+
+        # extract and rescale means, mixture componenets and covariances
+        nn = self.net
+        dist = nn._distribution
+
+        if context == None:
+            encoded_x = nn._embedding_net(self.default_x)
+        else:
+            encoded_x = nn._embedding_net(torch.tensor(context, dtype=torch.float32))
+
+        logits, m, prec, sumlogdiag, _ = dist.get_mixture_components(encoded_x)
+        norm_logits = logits - torch.logsumexp(logits, dim=-1, keepdim=True)
+
+        scale = nn._transform._transforms[0]._scale
+        shift = nn._transform._transforms[0]._shift
+
+        means_transformed = ((m - shift) / scale).detach()
+        
+        A = scale * torch.eye(means_transformed.shape[2])
+        precision_factors_transformed = torch.cholesky(A@prec@A)
+        
+        self.logits = norm_logits.detach()
+        self.means = means_transformed.detach()
+        self.precs = precision_factors_transformed.detach()
+        self.sumlogdiag = torch.sum(torch.log(torch.diagonal(self.precs, dim1=2, dim2=3)),dim=2).detach()
+        
+        return norm_logits, means_transformed, precision_factors_transformed, sumlogdiag
 
     @staticmethod
     def mulnormpdf(X: Tensor, mu: Tensor, cov: Tensor) -> Tensor:
```

Comment:
Is there an implementation of a multivariate normal in sbi somewhere that I am not aware of? 
I could also use `torch.distributions.MultivariateNormal`, but then I create objects in a loop again, although a small one.

Reply:
Maybe I am missing something, but: why do you need a multivariate normal? Isn't `log_prob_mog` sufficient for what you want to do?

Reply:
oh yeah.... I totally overlooked that, whoops :smile: